engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoContext.java

```diff
@@ -337,6 +337,7 @@ public Optional<Module> findModule(String moduleName) {
    */
   public Optional<Module> findModuleByExpressionId(UUID expressionId) {
     return getTopScope().getModules().stream()
+        .filter(module -> module.getIr() != null)
         .filter(
             module ->
```

Comment:
Two `.filter` statements one after another!? They could be merged into one and just `module.getIr() != null && ...`

Reply:
I thought Java will optimize those calls, but apparently it doesn't do that.

Reply:
Pesky side effects

Reply:
> I thought Java will optimize those calls, but apparently it doesn't do that.

The Graal compiler (from bytecode to assembly) might optimize it. However I've heard numerous discussions about how hard it is to optimize `stream`s. Btw. it is forbidden to use streams/lambdas in Truffle code - even if their peak performance is OK, they slow the initialization of JVM down.

As such I'd merge the two `filter` checks into one. Less burden on the JVM.

Reply:
I doubt JVM will optimize it

Reply:
FYI: [stream examples](https://www.graalvm.org/latest/examples/java-performance-examples/).